modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/pagemem/SpeedBasedThrottleIntegrationTest.java

```diff
@@ -0,0 +1,114 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.cache.persistence.pagemem;
+
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.ignite.Ignite;
+import org.apache.ignite.IgniteCache;
+import org.apache.ignite.configuration.DataRegionConfiguration;
+import org.apache.ignite.configuration.DataStorageConfiguration;
+import org.apache.ignite.configuration.IgniteConfiguration;
+import org.apache.ignite.internal.processors.cache.persistence.db.SlowCheckpointMetadataFileIOFactory;
+import org.apache.ignite.testframework.ListeningTestLogger;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.junit.Test;
+
+import static org.apache.ignite.cluster.ClusterState.ACTIVE;
+
+/**
+ * Integration tests for {@link PagesWriteSpeedBasedThrottle}.
+ */
+public class SpeedBasedThrottleIntegrationTest extends GridCommonAbstractTest {
+    /***/
+    private final ListeningTestLogger listeningLog = new ListeningTestLogger(log);
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String gridName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(gridName);
+
+        DataStorageConfiguration dbCfg = new DataStorageConfiguration()
+            .setDefaultDataRegionConfiguration(new DataRegionConfiguration()
+                // set small region size to make it easy achieve the necessity to throttle with speed-based throttle
+                .setMaxSize(60 * 1024 * 1024)
+                .setPersistenceEnabled(true)
+            )
+            .setCheckpointFrequency(200)
+            .setWriteThrottlingEnabled(true)
+            .setFileIOFactory(
+                new SlowCheckpointMetadataFileIOFactory(
+                    new AtomicBoolean(true), TimeUnit.MILLISECONDS.toNanos(10000)
+                )
+            );
+
+        return cfg.setDataStorageConfiguration(dbCfg)
+            .setConsistentId(gridName)
+            .setGridLogger(listeningLog);
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        stopAllGrids();
+
+        super.beforeTest();
+
+        cleanPersistenceDir();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        stopAllGrids();
+
+        super.afterTest();
+
+        cleanPersistenceDir();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected long getTestTimeout() {
+        return 3 * 60 * 1000;
+    }
+
+    /**
+     */
+    @Test
+    public void speedBasedThrottleShouldBeActivatedWhenNeeded() throws Exception {
+        AtomicBoolean throttled = new AtomicBoolean(false);
+        listeningLog.registerListener(message -> {
```

Comment:
There are regexp-based or substring-based matchers for these loggers, you should probably use them. Just check out other usages

Reply:
The existing library leads to a code that is not too comprehensible. Like ` if (logListener.check()) { break; }` - it would make a reader wonder what kind of a check is it? If the method would be named appropriately, like `expectationsSatisfied()`, it would make sense to use it. With the current state of the code, it seems that it's easier to understand the code I wrote than use the 'standard' tools.